### PR TITLE
Add: enable command  line command

### PR DESCRIPTION
**Pull request by Jim8y**
This pr add `commandline` commands to neo-cli:

commands:

```C#
new Option<string>(new[] { "-c", "--config","/config" }, "Specifies the config file."),
new Option<string>(new[] { "-w", "--wallet","/wallet" }, "The path of the neo3 wallet [*.json]."),
new Option<string>(new[] { "-p", "--password" ,"/password" }, "Password to decrypt the wallet, either from the command line or config file."),
new Option<string>(new[] { "--db-engine","/db-engine" }, "Specify the db engine."),
new Option<string>(new[] { "--db-path","/db-path" }, "Specify the db path."),
new Option<string>(new[] { "--noverify","/noverify" }, "Indicates whether the blocks need to be verified when importing."),
new Option<string[]>(new[] { "--plugins","/plugins" }, "The names of the plugins [plugin1 plugin2]."),
```

### CODE REVIEW

**Comment by cschuchardt88** [comment]
This is starting to turn into `neo-express`? What is the point of this?

**Comment by Jim8y** [comment]
> This is starting to turn into `neo-express`? What is the point of this?

Will this hurt?

File: src/Neo.CLI/Settings.cs [hunk 1]
```diff
@@ -44,24 +44,30 @@ public static Settings Default
                     Initialize(config);
                 }
 
-                return _default;
+                return Custom ?? _default;
             }
         }
 
+        public static Settings? Custom { get; set; }
```
Comment (shargon) [inline src/Neo.CLI/Settings.cs:51]: We should wait for https://github.com/neo-project/neo/pull/3042

Comment (Jim8y) [reply src/Neo.CLI/Settings.cs:51]: @shargon

Comment (shargon) [reply src/Neo.CLI/Settings.cs:51]: Now we need to fix the conflicts  🤣

Comment (Jim8y) [reply src/Neo.CLI/Settings.cs:51]: 🥹🥹🥹🥹🥹

File: src/Neo.CLI/CLI/MainService.CommandLine.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        private static void CustomProtocolSettings(CommandLineOptions options, ProtocolSettings settings)
+        {
+            uint network = settings.Network;
+            ProtocolSettings tempSetting = settings;
+            // if specified network without specifying config, then load the default config
+            if (!string.IsNullOrEmpty(options.Network))
+            {
+                if (!uint.TryParse(options.Network, out network))
+                {
+                    network = options.Network switch
```
Comment (shargon) [inline src/Neo.CLI/CLI/MainService.CommandLine.cs:54]: I prefer to only specify the config, this hardcode the values

Comment (Jim8y) [reply src/Neo.CLI/CLI/MainService.CommandLine.cs:54]: Mainnet id should never change. It is supposed to be hardcoded into the code.

**Review by shargon — COMMENTED** [review]
Why src/Neo.CLI/CLI/CommandLineOption.cs warnings if it was not changed and we already merge some commits without this problem

File: src/Neo.CLI/CLI/MainService.CommandLine.cs (hunk 2 above)
Comment (shargon) [reply src/Neo.CLI/CLI/MainService.CommandLine.cs:54]: > Mainnet id should never change. It is supposed to be hardcoded into the code.

But we already use config for it

File: src/Neo.CLI/Settings.cs [hunk 3]
```diff
@@ -44,25 +44,30 @@ public static Settings Default
                     var config = new ConfigurationBuilder().AddJsonFile("config.json", optional: true).Build();
                     Initialize(config);
                 }
-
-                return s_default!;
+                return Custom ?? s_default!;
             }
         }
 
+        public static Settings? Custom { get; set; }
```
Comment (shargon) [inline src/Neo.CLI/Settings.cs:51]: Set `s_default`?

**Review by shargon — DISMISSED** [review]
Any way to use /help with this package?

File: src/Neo.CLI/Settings.cs (hunk 3 above)
Comment (Jim8y) [reply src/Neo.CLI/Settings.cs:51]: no, its not a property.

Comment (cschuchardt88) [reply src/Neo.CLI/Settings.cs:51]: he means it a property and no need to change it @shargon

Comment (Jim8y) [reply src/Neo.CLI/Settings.cs:51]: > he means it a property and no need to change it @shargon

LOL, u r right

**Comment by shargon** [comment]
@Jim8y Conflicting files :S

**Comment by shargon** [comment]
You don't want to move the config files?

File: src/Neo.CLI/CLI/MainService.cs [hunk 4]
```diff
@@ -366,26 +367,22 @@ public void OpenWallet(string path, string password)
...
-
+            CustomProtocolSettings(options, protocol);
+            CustomApplicationSettings(options, Settings.Default);
             NeoSystem = new NeoSystem(protocol, Settings.Default.Storage.Engine, string.Format(Settings.Default.Storage.Path, protocol.Network.ToString("X8")));
             NeoSystem.AddService(this);
 
             LocalNode = NeoSystem.LocalNode.Ask<LocalNode>(new LocalNode.GetInstance()).Result;
 
+            // installing plugins
+            options.Plugins?.Select(p => p).Where(p => !string.IsNullOrEmpty(p)).ToList().ForEach(async p => await InstallPluginAsync(p));
```
Comment (shargon) [inline src/Neo.CLI/CLI/MainService.cs:384]: https://stackoverflow.com/a/18668733

Comment (Jim8y) [reply src/Neo.CLI/CLI/MainService.cs]: Done

**Review by shargon — DISMISSED** [review]

**Review by vncoelho — COMMENTED** [review]
```cs
/workspaces/neo/src/Neo.CLI/Settings.cs(65,16): warning CS8618: Non-nullable property 'Contracts' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/workspaces/neo/src/Neo.CLI/Neo.CLI.csproj::TargetFramework=net7.0]
/workspaces/neo/src/Neo.CLI/CLI/MainService.CommandLine.cs(79,21): error CS0200: Property or indexer 'StorageSettings.Engine' cannot be assigned to -- it is read only [/workspaces/neo/src/Neo.CLI/Neo.CLI.csproj::TargetFramework=net7.0]
/workspaces/neo/src/Neo.CLI/CLI/MainService.CommandLine.cs(80,21): error CS0200: Property or indexer 'StorageSettings.Path' cannot be assigned to -- it is read only [/workspaces/neo/src/Neo.CLI/Neo.CLI.csproj::TargetFramework=net7.0]

Build FAILED.

/workspaces/neo/src/Neo.CLI/Settings.cs(65,16): warning CS8618: Non-nullable property 'Contracts' must contain a non-null value when exiting constructor. Consider declaring the property as nullable. [/workspaces/neo/src/Neo.CLI/Neo.CLI.csproj::TargetFramework=net7.0]
/workspaces/neo/src/Neo.CLI/CLI/MainService.CommandLine.cs(79,21): error CS0200: Property or indexer 'StorageSettings.Engine' cannot be assigned to -- it is read only [/workspaces/neo/src/Neo.CLI/Neo.CLI.csproj::TargetFramework=net7.0]
/workspaces/neo/src/Neo.CLI/CLI/MainService.CommandLine.cs(80,21): error CS0200: Property or indexer 'StorageSettings.Path' cannot be assigned to -- it is read only [/workspaces/neo/src/Neo.CLI/Neo.CLI.csproj::TargetFramework=net7.0]
    1 Warning(s)
    2 Error(s)
 ```

I think that I destroyed something when merged and conflicts resolved

**Comment by Jim8y** [comment]
I dont know, my local test passes.

File: src/Neo.CLI/Settings.cs [hunk 5]
```diff
@@ -60,32 +61,44 @@ public Settings(IConfigurationSection section)
             P2P = new(section.GetSection(nameof(P2P)));
             UnlockWallet = new(section.GetSection(nameof(UnlockWallet)));
         }
+
+        public Settings()
+        {
+            Logger = new LoggerSettings();
```
Comment (shargon) [inline src/Neo.CLI/Settings.cs:67]: Contracts must be initialized

File: src/Neo.CLI/CLI/MainService.cs [hunk 6]
```diff
@@ -366,26 +368,25 @@ public void OpenWallet(string path, string password)
...
-            bool verifyImport = true;
-            for (int i = 0; i < args.Length; i++)
-                switch (args[i])
-                {
-                    case "/noverify":
-                    case "--noverify":
-                        verifyImport = false;
-                        break;
-                }
+            bool verifyImport = options.NoVerify ?? true;
```
Comment (shargon) [inline src/Neo.CLI/CLI/MainService.cs:374]:
```suggestion
            bool verifyImport = !(options.NoVerify ?? false);
```

We equal NoVerify to verify, it must be the opposite, isn't it?

File: src/Neo.CLI/Settings.cs (hunk 5 above)
Comment (Jim8y) [reply src/Neo.CLI/Settings.cs:67]: This is not my code, it was here like this.

**Review by shargon — APPROVED** [review]

**Comment by cschuchardt88** [comment]
You have changed the interface. In doing so, i have found a bug. As you can see from the images below. That the `CLI` prompt is still open. When people try to run a command like `show state` it will clear the screen and render nothing. Please revert!!!

### Run `.\neo-cli.exe /noverify`
image

### when you type in `show state`
image



@shargon why did you merge this? You need two approvals. I only see one.

File: src/Neo.CLI/CLI/MainService.cs [hunk 7]
```diff
@@ -366,26 +368,25 @@ public void OpenWallet(string path, string password)
...
 
-        public async void Start(string[] args)
+        public async void Start(CommandLineOptions options)
         {
             if (NeoSystem != null) return;
-            bool verifyImport = true;
-            for (int i = 0; i < args.Length; i++)
-                switch (args[i])
-                {
-                    case "/noverify":
```
Comment (cschuchardt88) [inline src/Neo.CLI/CLI/MainService.cs:376]: This needs to be added back somehow.

**Comment by shargon** [comment]
> You have changed the interface. In doing so, i have found a bug. As you can see from the images below. That the `CLI` prompt is still open. When people try to run a command like `show state` it will clear the screen and render nothing. Please revert!!!
> 
> ### Run `.\neo-cli.exe /noverify`
> image
> 
> ### when you type in `show state`
> image
> 
> @shargon why did you merge this? You need two approvals. I only see one.

We can add in the readme the backwards incompatible changes, but we can't preserve all the  previous version if we try to improve the new one,  when it's released it could come with this information. With this change, the cli is much better, and people only need to change /verify to --xxx

**Comment by shargon** [comment]
> @shargon why did you merge this? You need two approvals. I only see one.

Is not my pr, so technically is @Jim8y and me, is two.

**Comment by cschuchardt88** [comment]
> > You have changed the interface. In doing so, i have found a bug. As you can see from the images below. That the `CLI` prompt is still open. When people try to run a command like `show state` it will clear the screen and render nothing. Please revert!!!
> > ### Run `.\neo-cli.exe /noverify`
> > image
> > ### when you type in `show state`
> > image
> > @shargon why did you merge this? You need two approvals. I only see one.
> 
> We can add in the readme the backwards incompatible changes, but we can't preserve all the previous version if we try to improve the new one, when it's released it could come with this information. With this change, the cli is much better, and people only need to change /verify to --xxx

that is fine. 

Can you fix the bug with the `cli` prompt bugging out on invalid arguments on command-line? `cli` command `show state` doesn't work anymore when that happens.

**Comment by shargon** [comment]
> > > You have changed the interface. In doing so, i have found a bug. As you can see from the images below. That the `CLI` prompt is still open. When people try to run a command like `show state` it will clear the screen and render nothing. Please revert!!!
> > > ### Run `.\neo-cli.exe /noverify`
> > > image
> > > ### when you type in `show state`
> > > image
> > > @shargon why did you merge this? You need two approvals. I only see one.
> > 
> > We can add in the readme the backwards incompatible changes, but we can't preserve all the previous version if we try to improve the new one, when it's released it could come with this information. With this change, the cli is much better, and people only need to change /verify to --xxx
> 
> that is fine. 
> 
> Can you fix the bug with the `cli` prompt bugging out on invalid arguments on command-line? `cli` command `show state` doesn't work anymore when that happens.

We found a bug, we fix the bug, we can open a new pr to fix it, please open a issue, remember that this version is not released yet.

**Pull request merged**